fix(metro-config): Add missing dedup-guards when adding namespace-imports for live-bindings

kitten opened the pull request:
# Why

There's two blocks that add a namespace-import template with the same local name in the same "block" handling exports, but fail to deduplicate when the shared import was already added.

This could be made less likely if we:
- Always add `var LOCAL = require(SOURCE);` and then mark `LOCAL` as used when we access it
- Share `sharedModuleExportFrom` and add it repeatedly (let it duplicate) then deduplicate by reference equality
- (Most heavy-handed) key import/exports by their source string

We could refactor this later but for now the missing guards are easily identified.

# How

- Add missing `!sharedModuleExportFrom` guards when adding `importAllTemplate` for `ExportNamedDeclaration` specifiers

# Test Plan

- Unit test added and addressed

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

kitten committed:
Add failing test for re-export conflict

kitten committed:
Add failing test for >1 locals after `default as` export specifier

kitten committed:
Add missing guards when adding export-all to avoid duplicates

kitten committed:
Rebuild

kitten committed:
Add CHANGELOG entry

kitten committed:
Add similar test for imports

douglowder reviewed (commented):
Looks good… I’ll test this locally in the repo that has d3-zoom

kitten committed:
Fix broken import-default reuse case

kitten committed:
Rebuild

krystofwoldrich reviewed (approved):
Thank you! Looks good.

kitten committed:
test(metro-config): Add compact snapshot cases for import-export-plugin (#38955)

Follow-up to #38951

We should have simple test cases listed with output snapshots that
attempt to cover every single possible case to have a good overview over
outputs.

The only omitted part here are to combine every case with a case of
another module, but that would explode into too many cases, and we
should simply trust that modules won't overlap or can't combine in the
transform.

## Failed cases

| Case | Note |
| -- | -- |
| `with live bindings transforms export import default specifier as
local 1` | Declares `var local` but live binding references
`_apple_icons` |
| `with live bindings transforms export import default specifier as
local, and default 1` | Declares `var local` but live binding references
`_apple_icons` |

---------

Co-authored-by: Krystof Woldrich <31292499+krystofwoldrich@users.noreply.github.com>
